### PR TITLE
Read dependencies from requirement files for setup

### PR DESCRIPTION
**Pull request by ankitade**
Summary:
- setup.py now has install_requires which reads from requirements.txt
- extra requires for dev specific dependencies coming from dev-requirements.txt (moved pytest to it)
- pytorch and torch* dependencies still need to be installed manually since we depend on nightly builds and there seems to be no principled way of getting them in setup.py

Test plan:
 - In a fresh conda env, followed Readme and tried to import flava. Checked no dev dependencies were installed in the env
 - In another env, followed updated contribution guide and ran pytest. Verified mypy, pytest and precommit were installed

### CODE REVIEW

File: dev-requirements.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+pytest
+mypy
+pre-commit
```
Comment (ebsmothers) [inline dev-requirements.txt:3]: Should we also add flake8/ufmt/black/usort to dev requirements?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -33,7 +33,7 @@ outlined on that page and do not file a public issue.
 
 Same as in [README](README.md) with the exception of:
 ```
-python setup.py develop
+pip install -e ".[dev]"
```
Comment (ebsmothers) [inline CONTRIBUTING.md:36]: We may also want to change the instructions in [README.md](https://github.com/facebookresearch/multimodal/blob/main/README.md#building-from-source)

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -77,13 +76,7 @@ ufmt format `git diff main --name-only`
...
 ### Type checking
 
-TorchMultimodal uses mypy for type checking. You can install mypy with the command
-
-```
-python3 -m pip install mypy
-```
-
-To perform type checking:
+TorchMultimodal uses mypy for type checking.To perform type checking:
```
Comment (langong347) [inline CONTRIBUTING.md:79]: Adding a space here?
`TorchMultimodal uses mypy for type checking.(space)To perform type checking:`

File: README.md [hunk 4]
```diff
@@ -30,7 +30,6 @@ TorchMultimodal requires Python >= 3.8. The library can be installed with or wit
     ```
     git clone --recursive https://github.com/facebookresearch/multimodal.git torchmultimodal
     cd torchmultimodal
-    pip install -r requirements.txt
 
     python setup.py install
```
Comment (langong347) [inline README.md:34]: Since we now read requirements into setuptools, we can `pip install torchmultimodal`?

**Comment by langong347** [comment]
We should consider adding version numbers in requirements file at least for packages where reproducible behaviors are important.

File: dev-requirements.txt (hunk 1 above)
Comment (ankitade) [reply dev-requirements.txt:3]: it should come from precommit right?

**Comment by ankitade** [comment]
this is pretty much reproducing the manual readme right now so idk which ones we want to pin

File: dev-requirements.txt (hunk 1 above)
Comment (ebsmothers) [reply dev-requirements.txt:3]: Yeah pre-commit will run these regardless, I am thinking more if developers want to run the checks themselves. But no strong preference here

Comment (langong347) [reply dev-requirements.txt:3]: pre-commit will run more than just linting and autoformatting including license header, AST and etc.
If the user wants to run linting/formatter separately, they will need to install flake8/umft/black/usort themselves.
The versions are also important to ensure consistency with fb internal. We documented the versions in contributing guide some time ago.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -33,7 +33,7 @@ outlined on that page and do not file a public issue.
 
 Same as in [README](README.md) with the exception of:
 ```
-python setup.py develop
+pip install -e .
```
Comment (langong347) [inline CONTRIBUTING.md:36]: Does this now replace the two-step approach `python setup.py [mode]` + `pip install -r [mode]-requirements`?

Should we consider sticking with one flavor in our documentation (in README.md it still tells the users to do `python setup.py install`)?

**Comment by langong347** [comment]
> this is pretty much reproducing the manual readme right now so idk which ones we want to pin

Versions are added in requirements files. All packages should have versions to guarantee reproducibility across developers and (especially) in CI.

But I've also seen repos not being strict with versions (i.e., only package names in requirements). That should be fine until we become more sensitive to versions.

For now I'm just referring to the linter/formatter (see my comments earlier) as those are known issues.

File: CONTRIBUTING.md (hunk 5 above)
Comment (ankitade) [reply CONTRIBUTING.md:36]: strange, i meant to change it in requirements.txt and not in contributing but maybe i wasnt paying attention

File: dev-requirements.txt (hunk 1 above)
Comment (ankitade) [reply dev-requirements.txt:3]: just added back the install instructions for ufmt etc in contributing

**Review by ebsmothers — APPROVED** [review]

**Pull request closed without merging**